docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,338 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These redirects are useful, but we should rename them to something more general,
+since they apply to all types of projects, not just Sphinx projects.
+
+Proposed names:
+
+- HTML URL to clean URL redirect (``file.html`` to ``file/``)
+- Clean URL to HTML URL redirect (``file/`` to ``file.html``)
+
+Other ideas to improve redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Run forced redirects before built-in redirects.
+  We currently run built-in redirects before forced redirects,
+  this is a problem when moving a whole project to a new domain.
+  For example, a forced redirect like ``/$rest``,
+  won't work for the root URL of the project,
+  since ``/`` will first redirect to ``/en/latest/``.
+
+  But shouldn't be a real problem, since users will still need to
+  handle the ``/en/latest/file/`` paths.
+
+- Run redirects on the edge.
+  Cloudflare allow us to create redirects on the edge,
+  but they have some limitations around the number of
+  redirect rules that can be created.
+
+  And they will be useful for forced exact redirects only,
+  since we can't match a redirect based on the response of the origin server.
+
+- Merge prefix redirects with exact redirects.
+  Prefix redirects are the same as exact redirects with a wildcard at the end.
+
+- Merge all redirects into a single type.
+  This may simplify the implementation,
+  but it will make it harder to explain the feature to users.
+  And to replace some redirects we need to implement some new features.
+
+- Placeholders.
+  I haven't seen users requesting this feature.
+  We can consider adding it in the future.
+  Maybe we can expose the current language and version as placeholders.
+
+- Replace ``$rest`` with ``*`` in the from_url.
+  This will be more consistent with other services,
+  but it will require users to re-learn the feature.
+
+- Per-protocol redirects.
+  We should push users to always use HTTPS.
+
+- Allow a prefix wildcard.
+  We currently only allow a suffix wildcard,
+  adding support for a prefix wildcard should be easy.
+  But do users need this feature?
+
+- Per-domain redirects.
+  The main problem that originated this request was that we were applying redirects on external domains,
+  if we stop doing that, there is no need for this feature.
+  We can also try to improve how our built-in redirects work
+  (specially our canonical domain redirect).
+
+Allow matching query arguments
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We can do this in two ways:
+
+- At the DB level with some restrictions.
+  If done at the DB level,
+  we would need to have a different field
+  with just the path, and other with the query arguments normalized and sorted.
+
+  For example, if we have a redirect with the value ``/foo?blue=1&yellow=2&red=3``,
+  if would be normalized in the DB as ``/foo`` and ``blue=1&red=3&yellow=2``.
+  This implies that the URL to be matched must have the exact same query arguments,
+  it can't have more or less.
```

Comment:
This can be solved with a JSONField on the db to save the query arguments. In that case, the order and number of the arguments won't matter.

Reply:
But matching can't be done with one query even using a JSONField either, matching will need to happen on the Python side.

Reply:
I'm not sure to understand your reply.

I'm saying that with a JSONField that stores the query arguments you can perform the match of the query arguments at DB level. Once you get those, you can perform the path/URL matching with Python.

Reply:
If the current URl is `?one&two&three`, we don't know which of those arguments we need to match in the rules. A rule can be `one`, or be `two`, or `one&three`.

Reply:
I think this is solvable by "matching the query arguments _from the rule_ into the DB field" -- no the other way around as you are saying. Example:

```python
.filter(queryargs__haskeys=rule.queryargs.keys())
```

https://docs.djangoproject.com/en/4.2/ref/contrib/postgres/fields/#has-keys

or we can use `contained_by` if we want to match specific `key, value` pairs:

```python
.filter(queryargs__contained_by=rule.queryargs)
```

https://docs.djangoproject.com/en/4.2/ref/contrib/postgres/fields/#std-fieldlookup-hstorefield.contained_by

Reply:
oh, didn't know about those queries, I have updated the doc to mention this approach.